code/operators/solution/operators.sol.cpp

```diff
@@ -6,88 +6,77 @@
 class Fraction {
  public:
   explicit Fraction(int i) : m_num(i), m_denom(1) {}
-  Fraction(int num, int denom) : m_num(num), m_denom(denom) {}
+  Fraction(int num, int denom) : m_num(num), m_denom(denom) {
+    normalize();
+  }
 
-  int num() const { return m_num; }
-  int denom() const { return m_denom; }
+  friend std::ostream& operator<<(std::ostream& os, Fraction const & f) {
+    os << f.m_num << "/" << f.m_denom;
+    return os;
+  }
 
-  void normalize() {
-    const int gcd = std::gcd(m_num, m_denom);
-    m_num /= gcd;
-    m_denom /= gcd;
+  friend Fraction & operator*=(Fraction & f, int i) {
+    f.m_num *= i;
+    f.normalize();
+    return f;
   }
 
-  Fraction& operator*=(int i) {
-    m_num *= i;
-    normalize();
-    return *this;
+  friend Fraction & operator*=(Fraction & f, Fraction const & o) {
+    f.m_num *= o.m_num;
+    f.m_denom *= o.m_denom;
+    f.normalize();
+    return f;
   }
 
-  Fraction& operator*=(Fraction f) {
-    m_num *= f.num();
-    m_denom *= f.denom();
-    normalize();
-    return *this;
+  friend Fraction & operator+=(Fraction & f, int i) {
+    f.m_num += i * f.m_denom;
+    return f;
   }
 
-  Fraction& operator+=(int i) {
-    m_num += i * m_denom;
-    return *this;
+  friend Fraction & operator+=(Fraction & f, Fraction o) {
+    f.m_num *= o.m_denom;
+    f.m_num += o.m_num * f.m_denom;
+    f.m_denom *= o.m_denom;
+    f.normalize();
+    return f;
   }
 
-  Fraction& operator+=(Fraction f) {
-    m_num *= f.denom();
-    m_num += f.num() * m_denom;
-    m_denom *= f.denom();
-    normalize();
-    return *this;
+  friend bool operator==(Fraction const & a, Fraction const & b) {
+    return a.m_num == b.m_num && a.m_denom == b.m_denom;
+  }
+
+  friend bool operator<(Fraction const & a, Fraction const & b) {
+    return a.m_num * b.m_denom < b.m_num * a.m_denom;
   }
 
  private:
+  void normalize() {
+    const int gcd = std::gcd(m_num, m_denom);
+    m_num /= gcd;
+    m_denom /= gcd;
+  }
+
   int m_num, m_denom;
 };
 
-std::ostream& operator<<(std::ostream& os, Fraction r) {
-  os << r.num() << "/" << r.denom();
-  return os;
-}
-
-Fraction operator*(Fraction r, int i) { return r *= i; }
-
-Fraction operator*(int i, Fraction r) { return r * i; }
-
-Fraction operator*(Fraction a, Fraction b) { return a *= b; }
+Fraction operator*(Fraction f, int i) { return f *= i; }
+Fraction operator*(int i, Fraction const & f) { return f * i; }
+Fraction operator*(Fraction a, Fraction const & b) { return a *= b; }
 
 Fraction operator+(Fraction r, int i) { return r += i; }
+Fraction operator+(int i, Fraction const & r) { return r + i; }
+Fraction operator+(Fraction a, Fraction const & b) { return a += b; }
```

Comment:
These should be hidden friends.

Reply:
Here you define operators as free function. This adds them visibly to the global namespace. So, e.g. whenever you add two `std::string` like: `std::string a, b; a + b;`, the compiler will have to try your operators from `Fraction`  as well (and discard them). This causes a compile-time slowdown, because the overload set for `operator+` is unnecessary large. Here is a blog post with a synthetic benchmark claiming a 23x (!) reduction in compilation time using hidden friends: https://jacquesheunis.com/post/hidden-friend-compilation/

Secondly, if you now defined a class `Complex`, that had an implicit conversion operator to `Fraction`, your `Complex` would use all the operators from `Fraction`, because the compiler would just figure out the right conversion sequence. So a call to `Complex a, b; a + b` would be translated to: `Complex a, b; operator+()(Fraction(a), Fraction(b))`, which does not sound too crazy now, but can actually lead to big surprises. Again, hidden friends would not allow such a conversion because ADL could not find the `operator+` inside `Fraction` for two `Complex` arguments. More in this blob post: https://www.justsoftwaresolutions.co.uk/cplusplus/hidden-friends.html